Detectors/EMCAL/simulation/include/EMCALSimulation/DMAOutputStream.h

```diff
@@ -0,0 +1,117 @@
+
+#include <exception>
+#include <fstream>
+#include <string>
+
+#include <gsl/span>
+
+#include "Rtypes.h"
+#include "RStringView.h"
+
+#include "Headers/RAWDataHeader.h"
+
+namespace o2
+{
+
+namespace emcal
+{
+
+/// \class DMAOutputStream
+/// \brief Output stream of a payload to DMA raw files
+/// \author Markus Fasel <markus.fasel@cern.ch>, Oak Ridge National Laboratory
+/// \since Nov 6, 2019
+///
+/// Stream of output payload with variable size to DMA pages in
+/// a binary raw file. The output payload can be larger than the
+/// size of a DMA page (default: 8 kB) the output is split into
+/// multiple pages. Page counter, memory size, page size and
+/// stop bit are handled internally and are overwritten in the
+/// raw data header provided. All other header information must be
+/// handled externally.
+class DMAOutputStream
+{
+ public:
+  using RawHeader = o2::header::RAWDataHeaderV4;
+
+  class OutputFileException : public std::exception
+  {
+   public:
+    OutputFileException() = default;
+    OutputFileException(const std::string_view filename) : std::exception(), mFilePath(filename), mMessage("Path \"" + mFilePath + "\" invalid") {}
+    ~OutputFileException() noexcept final = default;
+
+    const char* what() const noexcept final
+    {
+      return mMessage.data();
+    }
+
+   private:
+    std::string mFilePath = "";
+    std::string mMessage = "";
+  };
+
+  /// \brief Constructor
+  DMAOutputStream() = default;
+
+  /// \brief Constructor
+  /// \param filename Name of the output file
+  DMAOutputStream(const char* filename);
+
+  /// \brief Destructor
+  ///
+  /// Closing file I/O
+  ~DMAOutputStream();
+
+  /// \brief Open the output stream
+  /// \throw OutputFileException
+  ///
+  /// Opening output file I/O
+  void open();
+
+  /// \brief Set the name of the output file
+  /// \param filename Name of the output file
+  void setOutputFilename(const char* filename) { mFilename = filename; }
+
+  /// \brief Write output payload to the output stream
+  /// \param header Raw data header
+  /// \param buffer Raw data payload
+  ///
+  /// Converting output payload to DMA papges. If the payload is larger than
+  /// the pagesize - header size the payload is automatically split to multiple
+  /// pages. Page counter, memory size, page size and stop bit of the header are
+  /// handled internally and are overwritten from the header provided. All other
+  /// header information must be provided externally.
+  void writeData(RawHeader header, gsl::span<char> buffer);
+
+ protected:
+  /// \brief Write DMA page
+  /// \param header Raw data header for page
+  /// \param payload Page payload (includes size of teh payload)
+  /// \param pagesize Size of the DMA page (not size of the payload)
+  ///
+  /// Expects that the size of the payload is smaller than the size ot the DMA
+  /// page. Parameter pagesize supporting variable page size.
+  void writeDMAPage(RawHeader header, gsl::span<char> payload, int pagesize);
```

Comment:
why don't you use ``const RawHeader& header`` here? The writeDMAPage just copies the external header to the buffer.